arm: stm32: Data cache (dcache) management update

lmajewski opened the pull request:
This patch series fixes problem with enabling at the same time STM32H7's Data cache and CONFIG_NOCACHE_MEMORY option.

On the STM32 it shall be possible to use memory buffers with `__nocache` attribute with data cache being enabled. 
This is not the best possible solution - one could fix cache coherency issues on the DMA driver level, but allows avoiding
performance penalty when data cache is disabled.

This solution has been tested on `nucleo_h743zi` with `littlefs` sample program.
`west build -p always -b nucleo_h743zi ./zephyr/samples/subsys/fs/littlefs -DCONF_FILE=prj_blk.conf`

This patch series comply with point '2.1' of https://github.com/zephyrproject-rtos/zephyr/issues/36471

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/45739

Commit:
kconfig: Add CONFIG_DCACHE option

This option is by default defined and explicitly enables the data
cache on a target platform.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
arm: stm32h7: Enable DCache when CONFIG_DCACHE is defined

Up till now the usage of CONFIG_NOCACHE_MEMORY also explicitly disables
data cache on the STM32H7 SoC.

With this change the usage of CONFIG_NOCACHE_MEMORY has been decoupled
from data cache enabling as new Kconfig option - namely
CONFIG_DCACHE is now used to explicitly enable it.

After this change it would be possible to use data cache on STM32H7
with DMA buffers, fragile to cache coherency issues, defined with
'__nocache' attribute.

Such approach would improve the overall STM32H7 performance until the
moment when proper (i.e. in-DMA) buffer cache management is developed.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

de-nordic commented on soc/arm/st_stm32/stm32h7/soc_m7.c:
```diff
@@ -62,12 +62,12 @@ static int stm32h7_init(const struct device *arg)
-#ifndef CONFIG_NOCACHE_MEMORY
-	if (!(SCB->CCR & SCB_CCR_DC_Msk)) {
-		SCB_EnableDCache();
+	if (!IS_ENABLED(CONFIG_DCACHE_DISABLE)) {
```
`IS_ENABLE(*_DISABLE)` look clunky. Maybe `DCACHE_DISABLE` should be renamed, because it is strange to check if something is enabled to know that it is disabled.

lmajewski replied:
I think that `CONFIG_DCACHE_DISABLE` naming is correct here. If it is set, then the cache shall be **explicitly** disabled . By default `DCache` is enabled most of the time, so introducing `CONFIG_DCACHE_ENABLED` with `default y` seems to me like an overkill. 

The problem is with the current code, where there is `#ifndef CONFIG_NOCACHE_MEMORY`
Maybe below code would be better?

``` 
if (IS_ENABLED(CONFIG_DCACHE_DISABLE)) {
    SCB_DisableDCache();
} else {
    if (!(SCB->CCR & SCB_CCR_DC_Msk)) {
        SCB_EnableDCache();
    }
}
```

However, in this code we do disable already disabled `DCache` when `CONFIG_DCACHE_DISABLED` is defined.

erwango reviewed (changes requested):
I don't think we should go with this solution as this is globally the same as https://github.com/zephyrproject-rtos/zephyr/pull/35165 which was already rejected

lmajewski commented:
> I don't think we should go with this solution as this is globally the same as #35165 which was already rejected

@erwango The #35165 has been already accepted and pulled - the `#ifndef CONFIG_NOCACHE_MEMORY` is already present in `soc/arm/st_stm32/stm32h7/soc_m7.c`. 

And the current situation (as it is now in Zephyr repo's - `main` branch) is IMHO just wrong and reduces system performance considerably. Customers are **really** concerned about this.

With this patch we are going one small step forward - with having DCache enabled with DMA running with `__nocache` attribute for crucial buffers, which are really small. 

Some STM32 drivers do require defining the  `CONFIG_NOCACHE_MEMORY` and as a result, now the DCache is disabled for the whole system. 

This patch allows developers to use DCache and also check if their `__nocache` defined DMA buffers also are handled correctly. Until Zephyr gains proper handling of cache coherency, this would reduce the possible performance degradation (as it is **now** noticeable by some Zephyr users).

erwango commented:
> @erwango The #35165 has been already accepted and pulled - the `#ifndef CONFIG_NOCACHE_MEMORY` is already present in `soc/arm/st_stm32/stm32h7/soc_m7.c`.

Ok, this is not strictly exact, but in the end you're right. 
#35165, which was affecting F7 was indeed rejected, but some time after, #37716, targeting H7 and doing the same exact thing was merged.
So, I do agree a change is required to offer possibility to use DCache and nocache memory sections.
But then, I think `DCACHE_DISABLE` should be `default n` and enabled only when required by drivers (like uart async api, cf `tests/drivers/uart/uart_async_api/boards/nucleo_h723zg.conf`)
Also, I'd suggest to extend to other M7 based cores, at least on STM32, ie STM32F7.

lmajewski commented:
> > @erwango The #35165 has been already accepted and pulled - the `#ifndef CONFIG_NOCACHE_MEMORY` is already present in `soc/arm/st_stm32/stm32h7/soc_m7.c`.
> 
> Ok, this is not strictly exact, but in the end you're right. #35165, which was affecting F7 was indeed rejected, but some time after, #37716, targeting H7 and doing the same exact thing was merged. So, I do agree a change is required to offer possibility to use DCache and nocache memory sections. But then, I think `DCACHE_DISABLE` should be `default n` and enabled only when required by drivers (like uart async api, cf `tests/drivers/uart/uart_async_api/boards/nucleo_h723zg.conf`) Also, I'd suggest to extend to other M7 based cores, at least on STM32, ie STM32F7.

The problem here is that we already have many systems with DCache disabled when `CONFIG_NOCACHE_MEMORY` is enabled. This patch series tries to **not** introduce regressions as much as possible.
It would be the system integrator responsibility to enable the DCACHE explicitly (i.e. `CONFIG_DCACHE_DISABLE=n`) when he/she is also planning to use nocache memory DMA buffers (as it is with mine nucleo board).

However, I don't mind to follow your guidelines if other developers (i.e. @de-nordic @Laczen @FRASTM @ABOSTM @nvlsianpu ) agree about this approach.

lmajewski commented:
> > > @erwango The #35165 has been already accepted and pulled - the `#ifndef CONFIG_NOCACHE_MEMORY` is already present in `soc/arm/st_stm32/stm32h7/soc_m7.c`.
> > 
> > 
> > Ok, this is not strictly exact, but in the end you're right. #35165, which was affecting F7 was indeed rejected, but some time after, #37716, targeting H7 and doing the same exact thing was merged. So, I do agree a change is required to offer possibility to use DCache and nocache memory sections. But then, I think `DCACHE_DISABLE` should be `default n` and enabled only when required by drivers (like uart async api, cf `tests/drivers/uart/uart_async_api/boards/nucleo_h723zg.conf`) Also, I'd suggest to extend to other M7 based cores, at least on STM32, ie STM32F7.
> 
> The problem here is that we already have many systems with DCache disabled when `CONFIG_NOCACHE_MEMORY` is enabled. This patch series tries to **not** introduce regressions as much as possible. It would be the system integrator responsibility to enable the DCACHE explicitly (i.e. `CONFIG_DCACHE_DISABLE=n`) when he/she is also planning to use nocache memory DMA buffers (as it is with mine nucleo board).
> 
> However, I don't mind to follow your guidelines if other developers (i.e. @de-nordic @Laczen @FRASTM @ABOSTM @nvlsianpu ) agree about this approach.

Gentle PING on this patch set ....

erwango commented:
> However, I don't mind to follow your guidelines if other developers (i.e. @de-nordic @Laczen @FRASTM @ABOSTM @nvlsianpu ) agree about this approach.

I'd go for deactivating it by default and add a note in the release note to inform users.

de-nordic commented:
> > However, I don't mind to follow your guidelines if other developers (i.e. @de-nordic @Laczen @FRASTM @ABOSTM @nvlsianpu ) agree about this approach.
> 
> I'd go for deactivating it by default and add a note in the release note to inform users.

@erwango 
So you want to go with `DCACHE_DISABLED=n` by default?

Shouldn't, also, the option be dependent on `CACHE_MANAGEMENT` and disable `DCACHE_LINE_SIZE_DETECT` and `DCACHE_LINE_SIZE` when `DCACHE_DISABLED=n`?

Wouldn't we maybe go further add something like 'HAS_HW_DCACHE', that would be selected by platform and that would then show the `DCACHE_*` options?

Maybe we can, instead of `DCACHE_DISABLED=n`, have `DCACHE=y` defined like
```
config DCACHE
    bool ...
    default y if HAS_HW_DCACHE

```
And it would be `y` for platforms that define the `HAS_HW_DCACHE` (or whatever) and user projects could still override it to `n`,

lmajewski commented:
> > > However, I don't mind to follow your guidelines if other developers (i.e. @de-nordic @Laczen @FRASTM @ABOSTM @nvlsianpu ) agree about this approach.
> > 
> > 
> > I'd go for deactivating it by default and add a note in the release note to inform users.
> 
> @erwango So you want to go with `DCACHE_DISABLED=n` by default?
> 
> Shouldn't, also, the option be dependent on `CACHE_MANAGEMENT` and disable `DCACHE_LINE_SIZE_DETECT` and `DCACHE_LINE_SIZE` when `DCACHE_DISABLED=n`?
> 
> Wouldn't we maybe go further add something like 'HAS_HW_DCACHE', that would be selected by platform and that would then show the `DCACHE_*` options?
> 
> Maybe we can, instead of `DCACHE_DISABLED=n`, have `DCACHE=y` defined like
> 
> ```
> config DCACHE
>     bool ...
>     default y if HAS_HW_DCACHE
> ```
> 
> And it would be `y` for platforms that define the `HAS_HW_DCACHE` (or whatever) and user projects could still override it to `n`,

@de-nordic Is there any example of `HAS_HW_DCACHE` usage already present in Zephyr? Personally, I would opt for enabling the `DCACHE` by default and then, only when integrator finds issues, allow disabling it.

However, such approach could break some existing STM32H7 boards, as they run with DCACHE disabled when drivers with `__nocache` buffers are used.

erwango commented:
> However, such approach could break some existing STM32H7 boards, as they run with DCACHE disabled when drivers with __nocache buffers are used.

This is a good example of this non sense situation. This should be fixed. And on these ones, I don't think this should raise issues.
Where I expect issues is on drivers (well, more on applications using these drivers actually) that don't use __nocache buffers, but need some, like uart_async where DMA is used. 
For these, in a first step, we can replace in tree current `CONFIG_NOCACHE_MEMORY=y` by `CONFIG_DCACHE=n`.
For sure this will have impact on out of tree users for sure, but anyhow current way of handling this on STM32H7 is wrong and should be fixed one day.

lmajewski commented:
> > However, such approach could break some existing STM32H7 boards, as they run with DCACHE disabled when drivers with __nocache buffers are used.
> 
> This is a good example of this non sense situation. This should be fixed. And on these ones, I don't think this should raise issues. Where I expect issues is on drivers (well, more on applications using these drivers actually) that don't use __nocache buffers, but need some, like uart_async where DMA is used. For these, in a first step, we can replace in tree current `CONFIG_NOCACHE_MEMORY=y` by `CONFIG_DCACHE=n`. For sure this will have impact on out of tree users for sure, but anyhow current way of handling this on STM32H7 is wrong and should be fixed one day.

Please correct my understanding if it is wrong:
- I need to introduce `CONFIG_DCACHE` for `STM32H7` (similar to patches sent with this PR) set by default to `y`
- All instances of `CONFIG_NOCACHE_MEMORY=y` shall be replaced by `CONFIG_DCACHE=n`
- We do accept out-of-tree boards breakage - boards maintainers' are responsible for checking if those will run with `DCACHE` enabled. 
- Support for `HAS_HW_DCACHE` is for now skipped

Or am I missing something?

erwango commented:
> Please correct my understanding if it is wrong:
> 
> * I need to introduce `CONFIG_DCACHE` for `STM32H7` (similar to patches sent with this PR) set by default to `y`

Yes. Use `CONFIG_DCACHE` in STM32H7 `soc_m7.c` to protect code enabling DCache.

> * All instances of `CONFIG_NOCACHE_MEMORY=y` shall be replaced by `CONFIG_DCACHE=n`

Yes, for STM32H7 based boards only.

> * We do accept out-of-tree boards breakage - boards maintainers' are responsible for checking if those will run with `DCACHE` enabled.

 Unfortunately, I don't see way to generate a smart warning, so I can think of only 2 ways to help out of tree users:
- Add note in `doc/releases/release-notes-3.1.rst`
- Send a mail to users/devel (I can do it if required).

> * Support for `HAS_HW_DCACHE` is for now skipped

Fine with me. Maybe @de-nordic can comment.

> Or am I missing something?

We'll aslo need to get STM32F7 to also use DCACHE, but that's another story.

lmajewski commented:
> > Please correct my understanding if it is wrong:
> > 
> > * I need to introduce `CONFIG_DCACHE` for `STM32H7` (similar to patches sent with this PR) set by default to `y`
> 
> Yes. Use `CONFIG_DCACHE` in STM32H7 `soc_m7.c` to protect code enabling DCache.
> 
> > * All instances of `CONFIG_NOCACHE_MEMORY=y` shall be replaced by `CONFIG_DCACHE=n`
> 
> Yes, for STM32H7 based boards only.
> 
> > * We do accept out-of-tree boards breakage - boards maintainers' are responsible for checking if those will run with `DCACHE` enabled.
> 
> Unfortunately, I don't see way to generate a smart warning, so I can think of only 2 ways to help out of tree users:
> 
>     * Add note in `doc/releases/release-notes-3.1.rst`
> 
>     * Send a mail to users/devel (I can do it if required).
> 
> 
> > * Support for `HAS_HW_DCACHE` is for now skipped
> 
> Fine with me. Maybe @de-nordic can comment.
> 
> > Or am I missing something?
> 
> We'll aslo need to get STM32F7 to also use DCACHE, but that's another story.

@de-nordic  Do you have any input or comments to the above plan?

Commit:
stm32: Add CONFIG_DCACHE=n to stm32h7 uart API tests

As now the CONFIG_NOCACHE_MEMORY is not responsible for controlling the
data cache on STM32H7 SoC, the CONFIG_DCACHE=n must be set explicitly
to preserve previous behavior as UART driver is not using no-cache
buffers.

Considering the above comment, the CONFIG_NOCACHE_MEMORY can be safely
removed.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
doc: Info about replacing usage of CONFIG_NOCACHE_MEMORY with CONFIG_DCACHE

Add information entry to release-notes-3.1.rst regarding data cache
management on the STM32H7 SoC.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
arm: Enable NOCACHE_MEMORY on nucleo_h743zi when littlefs sample is run

It has been validated that the nucleo_h743zi board works correctly,
with '__nocache' buffers and data caches enabled, so this patch enables
support for the former.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

lmajewski commented:
@erwango @de-nordic  Gentle ping on this patch set.

erwango reviewed (changes requested):
My bad, I had these comments pending, while I though they were already published.

lmajewski commented on tests/drivers/dma/chan_blen_transfer/boards/nucleo_h743zi.conf:
```diff
@@ -1,4 +1,4 @@
 CONFIG_DMA_TRANSFER_CHANNEL_NR_0=6
 CONFIG_DMA_TRANSFER_CHANNEL_NR_1=10
-CONFIG_NOCACHE_MEMORY=y
+CONFIG_DCACHE=n
```
Ok.

erwango replied:
Sorry if I was unclear.
Let's keep only `CONFIG_NOCACHE_MEMORY=y` as this is functional (test implements a non cacheable buffer) and more interesting than fully disabling the DCache.

erwango commented on tests/drivers/dma/loop_transfer/boards/nucleo_h743zi.conf:
```diff
@@ -1,3 +1,3 @@
-CONFIG_NOCACHE_MEMORY=y
+CONFIG_DCACHE=n
```
Sorry if I was unclear.
Let's keep only `CONFIG_NOCACHE_MEMORY=y` as this is functional (test implements a non cacheable buffer) and more interesting than fully disabling the DCache.

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_h743zi.conf:
```diff
@@ -1 +1 @@
-CONFIG_NOCACHE_MEMORY=y
+CONFIG_DCACHE=n
```
Sorry if I was unclear.
Let's keep only `CONFIG_NOCACHE_MEMORY=y` as this is functional (test implements a non cacheable buffer) and more interesting than fully disabling the DCache.

lmajewski replied:
But the `CONFIG_DCACHE=n` preserves the current behavior of those tests. Before this change the `CONFIG_NOCACHE_MEMORY=y` implicitly disabled D-Cache on `STM32H7`.

Please look on the following patch:
https://github.com/zephyrproject-rtos/zephyr/pull/44692/commits/60b215420e3759101c29b3c488cd88eea7b60751

(this is the next version of this patch set).

Leaving only `CONFIG_NOCACHE_MEMORY=y` in those tests will change the test setup, as for `STM32H7` D-Cache is now enabled by default, so we would test no-cache memory with enabled D-Cache.

IMHO, we shall remove `CONFIG_DCACHE=n` only when it has been tested and validated that D-Cache works correctly with `__nocache` buffers.

erwango replied:
> IMHO, we shall remove CONFIG_DCACHE=n only when it has been tested and validated that D-Cache works correctly with __nocache buffers.

I've tested it. Works ok

lmajewski replied:
Ok, so then I will drop the:
https://github.com/zephyrproject-rtos/zephyr/pull/44692/commits/60b215420e3759101c29b3c488cd88eea7b60751

patch and resend, OK?

lmajewski replied:
Ok. Thanks for pointing this out :-)

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_h743zi.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_NOCACHE_MEMORY=y
+CONFIG_DCACHE=n
```
You can remove this one. Test is ok with `CONFIG_NOCACHE_MEMORY=y`

erwango commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -71,6 +71,10 @@ Changes in this release
       CAN_MODE_LOOPBACK)``.
 
+* STM32H7 The `CONFIG_NOCACHE_MEMORY` no longer is responsible for disabling
+  data cache when defined. Now the newely introduced `CONFIG_DCACHE` explicitly
```
Maybe good specifying how to exactly use it, as this could be misleading.
```suggestion
  data cache when defined. Now the newly introduced `CONFIG_DCACHE=n` explicitly
```

erwango reviewed (changes requested):
One very last comment and we should be fine.

Now, since we're post release freeze, we should get this allowed to be merged in V3.1.0.
I'm creating a dedicated issue about this.

de-nordic commented:
> @de-nordic Is that ok to get this in V3.1.0 as this is globally a bug fix for STM32H7 platfom ?

Yes, it is fine by me, but I am not in charge of the release.

FRASTM reviewed (commented):
1. still a workaround to avoid cache coherency pb with Level1 cache memory when using DMA
2.  should the noDCACHE config be set as **depends on the STM32 DMA**  
3. To extend to stm32F7
(refer to [AN4839](https://www.st.com/content/ccc/resource/technical/document/application_note/group0/08/dd/25/9c/4d/83/43/12/DM00272913/files/DM00272913.pdf/jcr:content/translations/en.DM00272913.pdf))

lmajewski commented:
@erwango - Is there any easy way to restart the `Coding Guidelines / Run coding guidelines checks on patch series (PR) (pull_request)` task ? As it looks like the service was not available at the moment when the CI has been started?

And the second question - would it be possible to add this PR to `v3.1.0` ?

@FRASTM - Just a side note. In the document that you pointed out:
https://www.st.com/content/ccc/resource/technical/document/application_note/group0/08/dd/25/9c/4d/83/43/12/DM00272913/files/DM00272913.pdf/jcr:content/translations/en.DM00272913.pdf
**Mistakes to avoid and tips:** (point 4)

"Always better to use non-cacheable regions for DMA buffers. The software can use the MPU to set up a
non-cacheable memory block to use as a shared memory between the CPU and DMA."

It looks like even STM recommends using the `__nocache` attribute for DMA buffers, without the need to fine management of cache operations.

erwango commented:
> @erwango - Is there any easy way to restart the `Coding Guidelines / Run coding guidelines checks on patch series (PR) (pull_request)` task ? As it looks like the service was not available at the moment when the CI has been started?

I've restarted gh task.

> And the second question - would it be possible to add this PR to v3.1.0 ?

I'm trying to

FRASTM commented on tests/drivers/uart/uart_async_api/boards/nucleo_h743zi.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_NOCACHE_MEMORY=y
+CONFIG_DCACHE=n
```
is `CONFIG_NOCACHE_MEMORY=y` still required for this testcase ?

lmajewski replied:
Please see the above comment.

lmajewski commented on tests/drivers/uart/uart_async_api/boards/nucleo_h723zg.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_NOCACHE_MEMORY=y
```
As this driver (i.e. UART) is not using `no-cache` buffers - I do guess that `CONFIG_NOCACHE_MEMORY=y` was only used to disable data caches.
With this patch set applied - only `CONFIG_DCACHE=n` would be enough, so `CONFIG_NOCACHE_MEMORY=y`  could be removed.

@erwango - what do you think?

erwango replied:
> @erwango - what do you think?

Exact, good catch.

de-nordic commented on samples/subsys/fs/littlefs/boards/nucleo_h743zi.conf:
```diff
@@ -19,3 +19,5 @@ CONFIG_FS_LITTLEFS_PROG_SIZE=32
 CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE=64
 CONFIG_FS_LITTLEFS_FC_HEAP_SIZE=8192
+
+CONFIG_NOCACHE_MEMORY=y
```
Shouldn't this be `CONFIG_DCACHE=n`?

lmajewski replied:
No, this board with this test has been checked that works correctly.
